### PR TITLE
Hashed password Before storing it to db, Login route added

### PR DESCRIPTION
**Pull request by sounak07**
- Bcrypt js is used to hash.

- UserSchema.pre(a mongoose middleware) is to hash the password before storing it to db.

### CODE REVIEW

File: server/api/controllers/user.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ exports.user_signup = (req, res, next) => {
   }).then((token)=>{
     res.header("x-auth",token).send(user);
   }).catch((e) => {
-    res.status(400).send(e);
+    res.status(400).send(e.errmsg="User already exists");
```
Comment (skyshader) [inline server/api/controllers/user.js]: Can we just keep it `e.message` or `e.msg`?

File: server/api/models/user.js [hunk 2]
```diff
@@ -69,6 +70,21 @@ UserSchema.statics.findByToken = function(token){
...
+  var user = this;
+
+  if(user.isModified('password')){
+    bcrypt.genSalt(10,(err, salt)=>{
+      bcrypt.hash(user.password, salt, (err, hash)=>{
+        user.password = hash;
+        next();
+      });
+    });
+  } else {
```
Comment (skyshader) [inline server/api/models/user.js]: Let's early return here.

File: server/api/models/user.js [hunk 3]
```diff
@@ -69,6 +70,21 @@ UserSchema.statics.findByToken = function(token){
...
+    next();
+  }
+});
+
 var User = mongoose.model("User",UserSchema);
```
Comment (skyshader) [inline server/api/models/user.js]: Space after comma

Comment (skyshader) [reply server/api/models/user.js]: const instead of var

File: server/api/models/user.js [hunk 4]
```diff
@@ -69,6 +70,21 @@ UserSchema.statics.findByToken = function(token){
   });
 };
 
+UserSchema.pre("save",function(next){
+  var user = this;
```
Comment (skyshader) [inline server/api/models/user.js]: const instead of var

File: server/api/models/user.js [hunk 5]
```diff
@@ -2,6 +2,7 @@ const mongoose = require("mongoose");
 const validator = require("validator");
 const jwt = require("jsonwebtoken");
 const _ = require("lodash");
+const bcrypt = require("bcryptjs");
 
 
 var UserSchema = mongoose.Schema({
```
Comment (skyshader) [inline server/api/models/user.js]: const instead of var

File: server/api/models/user.js (hunk 3 above)
Comment (skyshader) [inline server/api/models/user.js]: const instead of var

**Comment by sounak07** [comment]
I have resolved all the issues but u also asked for compare password method but we need that while logging in a user, should write the login route as well along with all the changes..??

**Comment by skyshader** [comment]
Nope. Just keep a compare password method handy in the model. If you're hashing a password, there should also be a way to compare it.

**Comment by skyshader** [comment]
@sounak07 also, take care of the eslint violations because of previous PR :+1:

**Comment by sounak07** [comment]
Once I hash the password I don't have access to the original password as I am setting the user.password = hashed password
If I want to compare the password with hash I have to store the password in a variable before hashing but I don't know if that's a good practice... Please guide me on this @skyshader

**Comment by skyshader** [comment]
What you do typically is, given a password (which is not hashed) - hash it and compare with the hash stored in that user's model. But bcrypt gives a compare method to do the same.

Well, check this SO question: https://stackoverflow.com/questions/40076638/compare-passwords-bcryptjs

**Pull request merged**